contracts/LidoMatic.sol

```diff
@@ -0,0 +1,476 @@
+
+        uint256 amountInMATIC = getUserBalanceInMATIC();
+
+        sellVoucher_new(validatorShare, amountInMATIC, type(uint256).max);
+
+        if (validator2Nonce[validatorShare] == 0) {
+            validator2Nonce[validatorShare] = 1;
+        } else {
+            validator2Nonce[validatorShare]++;
+        }
+
+        if (user2Nonce[msg.sender] == 0) user2Nonce[msg.sender] = 1;
+
+        user2Nonce[msg.sender]++;
+
+        RequestWithdraw[] storage requestWithdraws = user2WithdrawRequest[
+            msg.sender
+        ];
+
+        requestWithdraws.push(
+            RequestWithdraw(
+                amountInMATIC,
+                validator2Nonce[validatorShare],
+                block.timestamp,
+                validatorShare,
+                true
+            )
+        );
+
+        lastWithdrawnValidatorId++;
+    }
+
+    /**
+     * @dev Delegates tokens to validator share contract
+     */
+    function delegate() external auth(DAO) {
+        Operator.OperatorShare[] memory operatorShares = nodeOperator
+            .getOperatorShares();
+
+        uint256 amountPerValidator = totalBuffered / operatorShares.length;
+        uint256 remainder = totalBuffered % operatorShares.length;
+
+        for (uint256 i = 0; i < operatorShares.length; i++) {
+            IERC20(token).approve(
+                operatorShares[i].validatorShare,
+                amountPerValidator
+            );
+
+            buyVoucher(operatorShares[i].validatorShare, amountPerValidator, 0);
+
+            validator2DelegatedAmount[
+                operatorShares[i].validatorShare
+            ] = amountPerValidator;
+        }
+
+        totalDelegated += totalBuffered - remainder;
+        totalBuffered = remainder;
+    }
+
+    /**
+     * @dev Claims tokens from validator share and sends them to the
+     * user if his request is in the userToWithdrawRequest
+     */
+    function claimTokens() external {
+        RequestWithdraw[] storage userRequests = user2WithdrawRequest[
+            msg.sender
+        ];
+
+        uint256 requestIndex;
+
+        for (uint256 i = userRequests.length - 1; i >= 0; i--) {
+            if (i > 0 && userRequests[i - 1].active) continue;
+            requestIndex = i;
+            break;
+        }
+
+        require(userRequests[requestIndex].active, "No active withdrawals");
+
+        require(
+            block.timestamp >=
+                userRequests[requestIndex].requestTime + WITHDRAWAL_DELAY,
+            "Not able to claim yet"
+        );
+
+        uint256 balanceBeforeClaim = IERC20(token).balanceOf(address(this));
+
+        unstakeClaimTokens_new(
+            userRequests[requestIndex].validatorAddress,
+            userRequests[requestIndex].validatorNonce
+        );
+
+        uint256 balanceAfterClaim = IERC20(token).balanceOf(address(this));
+        uint256 amount = balanceAfterClaim - balanceBeforeClaim;
+
+        IERC20(token).transfer(msg.sender, amount);
```

Comment:
1. put inside a require or safeTransfer
2. For the amount use the 
```suggestion
        IERC20(token).transfer(msg.sender, userRequests[id].amountInMATIC);
```
3. remove get balances